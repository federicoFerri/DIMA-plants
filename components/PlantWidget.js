import React from 'react';
import {Image, Text, SafeAreaView, TouchableOpacity, View } from 'react-native';
import firebase from "firebase";

class PlantWidget extends React.Component {
  state = { colorWaterStatus: null, timeLeftNextWatering: 0};

  componentDidMount() {
    const expired = firebase.firestore.Timestamp.now().seconds > this.props.plant.data().lastWatering.seconds + this.props.plant.data().secondsBetweenWaterings;
    const diff = this.props.plant.data().lastWatering.seconds + this.props.plant.data().secondsBetweenWaterings - firebase.firestore.Timestamp.now().seconds;
    console.log(firebase.firestore.Timestamp.now().seconds, this.props.plant.data().lastWatering.seconds + this.props.plant.data().secondsBetweenWaterings, diff, expired);
    this.setState({
        colorWaterStatus: (expired ? 'red' : 'green'),
        timeLeftNextWatering: (expired ? 'now' : 'in ' + Math.floor(diff / 60) + ' min')
    });
    // TODO start a recurrent operation (timer) that updates the time left and color of the icon (state)
  }

  componentWillUnmount() {
      // TODO: clear the interval
  }

    imagePressed = () => () => {
    this.props.onPress()
  }

  badPlantPressed = () => () => {
      firebase.firestore().collection('plants').doc(this.props.plant.id).update({
          logs: firebase.firestore.FieldValue.arrayUnion({date: firebase.firestore.Timestamp.now(), action: 'bad'}),
          secondsBetweenWaterings: this.props.plant.secondsBetweenWaterings + diff
      }).then(() => {
          // TODO: mark opacity with green color for n seconds and update props
      })
  }

  goodPlantPressed = () => () => {
      const diff = +10;
      firebase.firestore().collection('plants').doc(this.props.plant.id).update({
          logs: firebase.firestore.FieldValue.arrayUnion({date: firebase.firestore.Timestamp.now(), action: 'good'}),
          secondsBetweenWaterings: this.props.plant.secondsBetweenWaterings + diff
      }).then(() => {
          // TODO: mark opacity with green color for n seconds and update props
      })
  }

  wateringPlantPressed = () => () => {
      firebase.firestore().collection('plants').doc(this.props.plant.id).update({
          logs: firebase.firestore.FieldValue.arrayUnion({date: firebase.firestore.Timestamp.now(), action: 'watering'}),
          lastWatering: firebase.firestore.Timestamp.now()
      }).then(() => {
          // TODO: mark opacity with green color for n seconds and update props
      })
  }

  render() {
      return (
        <SafeAreaView style={{width: 300, height: 190, flex: 1, flexDirection:'column', justifyContent: 'center', alignItems: 'center', marginBottom: 20}}>
          <TouchableOpacity 
            activeOpacity={0.5}
            onPress={this.imagePressed()}>
                {/*plant image */}
                <Image 
                style={{width: 300, height: 140, opacity: 0.75, borderTopRightRadius: 10, borderTopLeftRadius: 10}}
                source={{uri: this.props.plant.data().imageUrl}}
                />
                {/*name of the plant */}
                <View style={{position: 'absolute', bottom: 2, left: 8}}>
                  <Text style={{fontSize: 25, color: '#000', fontFamily:'Comfortaa'}}>{this.props.plant.data().name}</Text>
                </View>
                {/*water status image */}
                <Image 
                style={{width: 20, height: 25, position: 'absolute', top:4, right:8}}
                tintColor={this.state.colorWaterStatus}
                source={require('../assets/button_images/water_status.png')}
                />
                {/*time left and clock image on top left */}
                <View style={{position: 'absolute', top: 2, left: 4, padding: 5, flex: 1, flexDirection:'row'}}>
                  <Image 
                  style={{width: 24, height: 23}}
                  source={require('../assets/button_images/clock.png')}
                  />
                  <Text style={{fontSize: 13, color: '#000', fontFamily: 'Comfortaa', padding: 2}}>next watering {this.state.timeLeftNextWatering}</Text>
                </View>
          </TouchableOpacity>
          <View style={{width: 300, padding: 10, flexDirection:'row', justifyContent: 'space-between', alignItems: 'center', backgroundColor: '#E8DFDF', borderBottomRightRadius: 10, borderBottomLeftRadius: 10}}>
            <TouchableOpacity activeOpacity={0.5} onPress={this.wateringPlantPressed()}>
                <Image 
                  style={{width: 30, height: 30}}
                  source={require('../assets/button_images/watering.png')}
                />
            </TouchableOpacity>
            <TouchableOpacity activeOpacity={0.5} onPress={this.goodPlantPressed()}>
                <Image 
                  style={{width: 30, height: 30}}
                  source={require('../assets/button_images/good_plant.png')}
                />
            </TouchableOpacity>
            <TouchableOpacity activeOpacity={0.5} onPress={this.badPlantPressed()}>
                <Image 
                  style={{width: 32, height: 30}}
                  source={require('../assets/button_images/bad_plant.png')}
                />
            </TouchableOpacity>
          </View>
        </SafeAreaView>
      );
  }
}

export default PlantWidget;