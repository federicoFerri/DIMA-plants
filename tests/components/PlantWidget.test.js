import React from 'react';
import renderer from 'react-test-renderer';

import PlantWidget from '../../components/PlantWidget'

describe('<PlantWidget />', () => {
    test('renders correctly', () => {
        const tree = renderer.create(<PlantWidget plant={{data: function(){return {lastWatering: {}}}}}
                                                  user={{}}
                                                  time_left_next_watering={30}
                                                  colorWaterStatus={'red'}
                                                  onPress={() => null}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});