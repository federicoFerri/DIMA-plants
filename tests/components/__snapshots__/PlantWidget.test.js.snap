// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlantWidget /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "height": 190,
      "justifyContent": "center",
      "marginBottom": 20,
      "width": 300,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Object {
          "borderTopLeftRadius": 10,
          "borderTopRightRadius": 10,
          "height": 140,
          "opacity": 0.75,
          "width": 300,
        }
      }
    />
    <View
      style={
        Object {
          "bottom": 2,
          "left": 8,
          "position": "absolute",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontFamily": "Comfortaa",
            "fontSize": 25,
          }
        }
      />
    </View>
    <Image
      source={1}
      style={
        Object {
          "height": 25,
          "position": "absolute",
          "right": 8,
          "tintColor": "green",
          "top": 4,
          "width": 20,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "left": 4,
          "padding": 5,
          "position": "absolute",
          "top": 2,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 23,
            "width": 24,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#000",
            "fontFamily": "Comfortaa",
            "fontSize": 11,
            "padding": 2,
          }
        }
      >
        next watering 
        now
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#E8DFDF",
        "borderBottomLeftRadius": 10,
        "borderBottomRightRadius": 10,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 10,
        "width": 300,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 30,
            "width": 32,
          }
        }
      />
    </View>
  </View>
</RCTSafeAreaView>
`;
